pkg/route/apiserver/admission/requiredrouteannotations/admission.go

```diff
@@ -0,0 +1,369 @@
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(pluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			klog.Infof("%s registered", pluginName)
+			return NewRequiredRouteAnnotations()
+		})
+}
+
+type requiredRouteAnnotations struct {
+	*admission.Handler
+	routeLister       routev1listers.RouteLister
+	nsLister          corev1listers.NamespaceLister
+	ingressLister     configv1listers.IngressLister
+	cachesSynced      bool
+	cachesToSync      []cache.InformerSynced
+	cachesToSyncNames []string
+	syncCachesOnce    sync.Once
+}
+
+// Ensure that the required OpenShift admission interfaces are implemented.
+var _ = admission.ValidationInterface(&requiredRouteAnnotations{})
+var _ = openshiftapiserveradmission.WantsOpenShiftConfigInformers(&requiredRouteAnnotations{})
+var _ = openshiftapiserveradmission.WantsOpenShiftRouteInformers(&requiredRouteAnnotations{})
+
+var maxAgeRegExp = regexp.MustCompile(`max-age=(\d+)`)
+
+// Validate ensures that routes specify required annotations, and returns nil if valid.
+func (o *requiredRouteAnnotations) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) (err error) {
+	if a.GetResource().GroupResource() != routev1.Resource("routes") {
+		return nil
+	}
+	newObject, isRoute := a.GetObject().(*routeapi.Route)
+	if !isRoute {
+		return nil
+	}
+
+	// Skip the validation if we're not updating/creating
+	switch a.GetOperation() {
+	case admission.Update, admission.Create:
+	default:
+		klog.Infof("ignoring %v on %s/%s", a.GetOperation(), a.GetNamespace(), a.GetName())
+		return nil
+	}
+
+	// Determine if there are HSTS changes in this update
+	if a.GetOperation() == admission.Update {
+		wants, has := false, false
+		var oldhsts, newhsts string
+
+		newmetadata, err := meta.Accessor(newObject)
+		if err != nil {
+			return err
+		}
+		newannotations := newmetadata.GetAnnotations()
+		if newannotations != nil {
+			newhsts, wants = newannotations[hstsAnnotation]
+		}
+
+		oldObject := a.GetOldObject()
+		if oldObject != nil {
+			oldmetadata, err := meta.Accessor(oldObject)
+			if err != nil {
+				return err
+			}
+			oldannotations := oldmetadata.GetAnnotations()
+			if oldannotations != nil {
+				oldhsts, has = oldannotations[hstsAnnotation]
+			}
+		}
+
+		// Skip the validation if we're not making a change to HSTS  this  update
+		if wants == has && newhsts == oldhsts {
+			klog.Infof("ignoring %v on %s/%s that doesn't change HSTS policy (old=%v, new=%v)", a.GetOperation(), a.GetNamespace(), a.GetName(), oldhsts, newhsts)
+			return nil
+		}
+	}
+
+	// Wait for all caches to sync at least once
+	o.syncCachesOnce.Do(func() {
+		o.cachesSynced = o.waitForSyncedStore(ctx)
+	})
+	if !o.cachesSynced {
+		synced := []string{}
+		for _, n := range o.cachesToSyncNames {
+			synced = append(synced, fmt.Sprintf("%s", n))
+		}
+		return admission.NewForbidden(a, errors.New(pluginName+": caches not synchronized: "+strings.Join(synced, ", ")))
+	}
```

Comment:
We should keep checking until it _succeeds_ once.
```suggestion
	if !o.syncCachesSynced {
		o.cachesSynced = o.waitForSyncedStore(ctx)
	}
	if !o.syncCachesSynced {
		return admission.NewForbidden(a, errors.New(pluginName+": caches not synchronized: "))
	}
	// all good, continue....
```
Nit: From what I've seen in the logs, we can tell which caches did not sync from the logs output by the informers.

Reply:
https://github.com/openshift/openshift-apiserver/pull/224/files#diff-cde26a88c8aa9772d3d1cbf96497ac0009504bc258fd662e2ad19b458dcd3d81R37

We are waiting for 10 seconds for it to succeed, but we don't want to wait forever in `Validate`.  Is that acceptable?

Reply:
Spoke with @frobware and I understand now.  Will post a change.